src/main/scala/chisel3/util/Counter.scala

```diff
@@ -19,27 +19,50 @@ import chisel3.internal.naming.chiselName  // can't use chisel3_ version because
   *     ...
   *   }
   * }}}
-  *
-  * @param n number of counts before the counter resets (or one more than the
-  * maximum output value of the counter), need not be a power of two
   */
 @chiselName
-class Counter(val n: Int) {
-  require(n >= 0, s"Counter value must be nonnegative, got: $n")
-  val value = if (n > 1) RegInit(0.U(log2Ceil(n).W)) else 0.U
+class Counter private (r: Range) {
+  require(r.start >= 0 && r.end >= 0, s"Counter range must be positive, got: $r")
+
+  private val delta = math.abs(r.step)
+  private val width = math.max(log2Up(r.last + 1), log2Up(r.head + 1))
 
-  /** Increment the counter
+  /** Creates a counter with the specified number of steps.
+    *
+    * @param n number of steps before the counter resets
+    */
+  def this(n: Int) { this(0 until n) }
+
+  /** The current value of the counter. */
+  val value = if (r.length > 1) RegInit(r.head.U(width.W)) else r.head.U
+
+  /** The range of the counter values. */
+  def range: Range = r
+
+  /** Increments the counter by a step.
     *
     * @note The incremented value is registered and will be visible on the next clock cycle
-    * @return whether the counter will wrap to zero on the next cycle
+    * @return whether the counter will wrap on the next clock cycle
     */
   def inc(): Bool = {
-    if (n > 1) {
-      val wrap = value === (n-1).U
-      value := value + 1.U
-      if (!isPow2(n)) {
-        when (wrap) { value := 0.U }
+    if (r.length > 1) {
+      val wrap = value === r.last.U
+
+      if (r.step > 0) {
+        // Increasing range
+        value := value + delta.U
+      } else {
+        // Decreasing range
+        value := value - delta.U
       }
+
+      // We only need to explicitly wrap counters that don't start at zero, or
+      // end on a power of two. Otherwise we just let the counter overflow
+      // naturally to avoid wasting an extra mux.
+      if (!(r.head == 0 && isPow2(r.last + delta))) {
```

Comment:
Thanks for addressing my concern.  I agree this works to cover all the old cases, as well as some new ones.  I do think there's a superset of this set that doesn't need to be reset, though: `isPow2((r.end - r.start).abs)`.  I don't think it matters that the initial value is 0, just that the total span is a power of 2.  Check my logic, though.

I'll let @jackkoenig review the rest.